---
output: 
  html_document:
    toc: yes
    toc_float:
      collapsed: false
pagetitle: Data Analytics Final Project
title: "Data Analytics Final Project - Constructing a Viral Phylogenetic Tree"
---
I know I don't have a whole ton of 'actual code' to show for this but I have been doing a lot of work figuring out what packages to use and how to use them.... lots and lots of reading!! and there is so so sooo much more to come!

phylogenies are very important for viral studies especially concerning viral evolution of potentially life threatening diseases. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, error = FALSE)
#CRAN Packages
library(dplyr)
library(tidyverse)
library(ape)
library(seqinr)
library(rentrez)
library(devtools)
library(stringr)



#Bioconductor packages 
library(msa)
library(Biostrings)


#github packages
library(compbio4all)
library(ggmsa)
```


## A download of the fasta file i created (150 complete viral genomes)
[complete list of virus genomes](http://megfrisby4.github.io/Data_Analytics_Project/Data/virus_genomes.FASTA)
-warning...it's a big old honker (2.1MB) beware!

## What I have so far.... (my exploratory code...definetly unfinished)
a few packages I need to construct my phylogeny (and acquire the correct data from NCBI):
```{r, message=FALSE}
#CRAN Packages
library(ape)
library(seqinr)
library(rentrez)
library(devtools)

#Bioconductor packages 
library(msa)
library(Biostrings)

#github packages
library(compbio4all)
library(ggmsa)

```

## Data 
The data table of many tears!

```{r}
   readr::read_csv("./Data/Virus_species_and_accession_data.csv") %>% kableExtra::kable() %>% kableExtra::kable_classic(lightable_options = 'hover') %>% kableExtra::scroll_box(width="500px", height = "200px")
```


## Using rentrez to explore sequences from NCBI nucleotide database 

some initial setup 
db = "nucleotide" indicates that I will be using accessing NCBI's nucleotide database.
rettype = "fasta" indicates each nucleotide sequence will be returned to me in fasta format 
```{r}
db = "nucleotide"
rettype = "fasta"

```


Here I am just exploring some functions from the new packages I downloaded. The entrez_fetch allows me to retrieve full genome nucleotide sequences from the NCBI database according to a unique accession number assigned to a tested viral sample.
The cat function organizes the sequence line by line so it is a bit easier to look at. 
Here I recall the full genomic sequence for Lloviu Cuevavirus and organize it in an easy to read format using cat
The genome is incredibly long so I will hide the output 
```{r, message=FALSE, results='hide'}
Lloviu_cuevavirus<- rentrez::entrez_fetch(db = "nucleotide",
                            id = "NC_016144", 
                             rettype = "fasta")

cat(Lloviu_cuevavirus)
```

Here I recall all the genomes of viruses in the family filoviridae by accession number in fasta format and make a list of the sequences manually 
```{r}
Mengla_dianlovirus <- rentrez::entrez_fetch(db = "nucleotide",
                                          id = "NC_055510", 
                                          rettype = "fasta")

Bombali_ebolavirus <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_039345", 
                      rettype = "fasta")

Bundibugyo_ebolavirus <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_014373", 
                      rettype = "fasta")

Reston_ebolavirus <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_004161", 
                      rettype = "fasta")

Sudan_ebolavirus <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_006432", 
                      rettype = "fasta")

Tai_Forest_ebolavirus <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_014372", 
                      rettype = "fasta")


Zaire_ebolavirus <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_002549", 
                      rettype = "fasta")


Marburg_marburgvirus_1 <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_001608", 
                      rettype = "fasta")

Marburg_marburgvirus_2 <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_024781", 
                      rettype = "fasta")


Wenling_frogfish_filovirus <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_055175", 
                      rettype = "fasta")

Wenling_thamnaconus_septentrionalis <- rentrez::entrez_fetch(db = "nucleotide",
                      id = "NC_055176", 
                      rettype = "fasta")

Bombali_ebolavirus <- fasta_cleaner(Bombali_ebolavirus, parse=F)

Bundibugyo_ebolavirus <- fasta_cleaner(Bundibugyo_ebolavirus, parse=F)

Reston_ebolavirus <- fasta_cleaner(Reston_ebolavirus, parse=F)

Sudan_ebolavirus <- fasta_cleaner(Sudan_ebolavirus, parse=F)

Tai_Forest_ebolavirus <- fasta_cleaner(Tai_Forest_ebolavirus, parse=F)

Zaire_ebolavirus <- fasta_cleaner(Zaire_ebolavirus, parse=F)

Marburg_marburgvirus_1 <- fasta_cleaner(Marburg_marburgvirus_1 , parse=F)

Marburg_marburgvirus_2 <- fasta_cleaner(Marburg_marburgvirus_2, parse=F)

Wenling_frogfish_filovirus <- fasta_cleaner(Wenling_frogfish_filovirus, parse=F)

Wenling_thamnaconus_septentrionalis <- fasta_cleaner(Wenling_thamnaconus_septentrionalis, parse=F)

filo_list <- list(Bombali_ebolavirus=Bombali_ebolavirus,
Bundibugyo_ebolavirus=Bundibugyo_ebolavirus,
Reston_ebolavirus=Reston_ebolavirus,
Sudan_ebolavirus=Sudan_ebolavirus,
Tai_Forest_ebolavirus=Tai_Forest_ebolavirus,
Zaire_ebolavirus=Zaire_ebolavirus,
Marburg_marburgvirus_1=Marburg_marburgvirus_1,
Marburg_marburgvirus_2=Marburg_marburgvirus_2,
Wenling_frogfish_filovirus=Wenling_frogfish_filovirus,
Wenling_thamnaconus_septentrionalis=Wenling_thamnaconus_septentrionalis)

```

I can access any sequence from the list like this:
filolist$[name of filo virus species]
Here's a sneak peek at the genome of the Bombali Ebolavirus:
```{r}
filo_list$Bombali_ebolavirus %>% glimpse()

```


## Exploring a bit of pairwise alignment
I can do a simple pairwise alignment with the Biostrings function pairwiseAlignment.
Here is the pairwise alignment of Bundibugyo and Reston Ebolaviruses:
```{r}
align_bundi_reston <- Biostrings::pairwiseAlignment(
 Bundibugyo_ebolavirus,
 Reston_ebolavirus)

align_bundi_reston

```

In general, the lower the score, the more  the genomes differ,  the higher the score the more similar the genomes are.
However, pairwise alignments scores are not very intuitive in and of themselves. 
Another function will tell us what percentage of the genomes align (pairwise sequence identity), a much more intuitive result!
```{r}
Biostrings::pid(align_bundi_reston)
```
All in all, these viruses share only 63.31% sequence identity! This goes to show how incredibly diverse viruses, even within the same family can be, and why they are so terrifying!

I predicted Bundibugyo ebolaviruse and Reston Ebolavirus would be quite different and they are in fact very diverse characteristically in both geographic origin and host range. 

Let's take a look at 2 viruses I believe will be more genetically similar, 2 different strains of Marburg Ebolavirus. 
```{r}
align_marburg_marburg <- Biostrings::pairwiseAlignment(
 Marburg_marburgvirus_1,
  Marburg_marburgvirus_2)

align_marburg_marburg

```

I can print just the alingment score alone with the score() function
```{r}
score(align_marburg_marburg)
```

A much higher score, but again, rather uninformative, and un-intuitive, lets have a look at the pid (percent identity)
```{r}
Biostrings::pid(align_marburg_marburg)
```
These genomes share about 79.3% sequence identity! They have much more in common than the previous two viruses, yet are still in the same family (taxonomic grouping)

## Pulling out the big guns: recalling and assembling 220 viral genomes 
```{r}
#remember i have filo from earlier (list of filoviridae accessions)
filo <- readr::read_lines("./Data/Filoviridae.acc_lst")
art <- readr::read_lines("./Data/Artoviridae.acc_lst")
born <- readr::read_lines("./Data/Bornaviridae.acc_lst")
lisp <- readr::read_lines("./Data/Lispiviridae.acc_lst")
mym <- readr::read_lines("./Data/Mymonaviridae.acc_lst")
nyan <- readr::read_lines("./Data/Nyamiviridae.acc_lst")
para <- readr::read_lines("./Data/Paramyxoviridae.acc_lst")
pneu <- readr::read_lines("./Data/Pneumoviridae .acc_lst")
sun <- readr::read_lines("./Data/Sunshinevirus.acc_lst")
xin <- readr::read_lines("./Data/Xinmoviridae.acc_lst")
rhab <- readr::read_lines("./Data/rhabdovirus.acc_lst")
virus_list <- c(filo,art,born,lisp,mym,nyan,para,pneu,sun,xin,rhab)
virus_list
```

I am going to load and utilize the BiomaRt package which helps my system to process and comprehend large amounts of data, then I will assemble a single list of complete genome sequences using my selection of accession numbers (corresponding to specific viruses) in fasta format.
I will accomplish this using the entrez_fetch_list() function from the campbio4all package. 
```{r, message=FALSE, results='hide'}
library(biomaRt)
virus_fasta_list <- entrez_fetch_list(db = "nucleotide", 
                                     id =virus_list, 
                                     rettype = "fasta")

#write.fasta(sequences = virus_fasta_list, names = names(virus_fasta_list), file.out = "./Data/virus_genomes.FASTA")


```
**Note**: access the FASTA I created of all 220 complete viral genomes with the link at the top of this webpage, you can view every single nucleotide for yourself! 

Creating a vector of family names that corresponds to the virus accession numbers.
Checking the length to make sure every genome is accounted for. 
Concatenating the accession number and family name into a data frame.
Saving as a csv for further display/use.
```{r}
s1 <- rep("Filoviridae", 12)
s2 <- rep("Artoviridae",7 )
s3 <- rep("Bornaviridae",18)
s4 <- rep("Lispiviridae", 6)
s5 <- rep("Mymonaviridae", 4)
s6 <- rep("Nyamiviridae", 13)
s7 <- rep("Paramyxoviridae",72)
s8 <- rep("Pneumoviridae", 9)
s9 <- rep("Sunshinevirus", 1)
s10 <- rep("Xinmoviridae",8)
s11 <- rep("Rhabdoviridae", 70)
fam <- c(s1,s2,s3,s4,s5,s6,s7,s8,s9,s10,s11)
length(fam)

df <- base::data.frame(accession_number=virus_list, virus_family=fam)
#write.csv(df, file="./Data/virus_acc_fam.csv")


```


## How I got my data table....
this was surprisingly hard but I harnessed my knowledge of lists and cleaning data and I did it and Im very happy because there was no way that I could just download this information.... that I knew of! So im proud I was able to use for loops and do this yay!

anyways heres the code for that:
```{r}
list_split <- list()
for (i in 1:length(virus_fasta_list)){list_split[[i]] <- str_split(virus_fasta_list[[i]],pattern = ',')
}

virus_names <- list()
for (i in 1:length(list_split)) {virus_names[[i]] <- purrr::map(list_split[[i]],1)
  
}


virus_species <- list()
virus_1 <- str_split(virus_names, "\\\\")
for (i in 1:length(virus_1)) {virus_species[[i]] <- purrr::map(virus_1[[i]][[1]],1)
  
}

virus_species <- unlist(virus_species)
virus_species <- str_remove(virus_species, "^...................")
virus_species <- str_remove(virus_species, "..$")

df <- data.frame(df, virus_species=virus_species)
write.csv(df, "./Data/Virus_species_and_accession_data.csv")

```


## Making all the sequences in my list useable! 

Cleaning the sequences into a neat formate I can perform alignments on!
Here is a peek:
```{r}
genome_list <- list()
for(i in 1:length(virus_fasta_list)){
  genome_list[[i]] <- fasta_cleaner(virus_fasta_list[[i]], parse = F)
}
glimpse(genome_list)
```


Making the list into a single vector of length 220 (one character string for each genome)
Im starting with an empty vector that I can then fill up, the output is literally a vector of 220 NA's:
```{r}
virus_vector <- rep(NA, length(genome_list))
virus_vector
```
Then I fill the  vector, replacing each NA with a full genome sequence.
As you can see the vector no longer cosistis of NA's, it now consists of DNA sequences. 
```{r}
for(i in 1:length(virus_vector)){
  virus_vector[i] <- genome_list[[i]]}
glimpse(virus_vector)
```
I will give the vector the virus names from what I extracted in the orginal fasta file previously 
```{r, results='hide'}
names(virus_vector) <- virus_species
```


## Converting to a stringset using BioStrings

Converting my newly made vector into a biostrings DNA stringset: 
```{r}
virus_vector_ss <- Biostrings::DNAStringSet(virus_vector)
virus_vector_ss 

```
mwhahahaha mad scientist moment!
**just a quick note**, most mononegavirales have rna genomes, as you can see, these sequences (and they are also this way in NCBI) represent DNA genomes, so keep in mind this would be the compliment to the actual viral genomes. The analyses will still work the same because these are all RNA viruses and so we are performing an analysis on all of their compliments, and their compliments are as equal to each other as their actual genomes are

## Performing Multiple sequence alignment 
How genomes align relative to eachother give scientists the information needed to construct a phylogenetic tree. An informative diagram of virus's evolutionary relationship to each other. 

phylogenetic trees are very important to pathology and entomology as they can help scientists predict properties of newly discovered viral species such as pathogenicity, zoonotic potential, vectorization method, and host range. 

```{r}
#virus_align <- msa(virus_vector_ss,
                     #method = "ClustalW")
```
This gives me the message 'use default substitution matrix' which tells me the program is working to come up with the ideal alignment of all the sequences by assigning values to matches mismatches and indels. 

now get a feel for what this new object is... specifically its class is 
```{r}

#class(virus_align)
```
By default the class of virus align is an msa object, I want the class to be just DNA multiple alignment so we'll need to do some minor tweaking. 
```{r, results='hide'}
#class(virus_align) <- "DNAMultipleAlignment"

```
I want to work with this object with functions from the seqinr package so Ill need to do a little more tweaking with a function called msaConvert
I will name the new object I am creating virus_align_seqinr so that I can get it all straight
the output is incredibly long there are 220 genomes here... so I will not show you.....youre welcome!
```{r, results='hide'}
#virus_align_seqinr <- msaConvert(virus_align, 
                                   #type = "seqinr::alignment")
```

looking at my full alignment 
```{r, results='hide'}
#glimpse(compbio4all::print_msa(alignment = virus_align_seqinr, 
          #chunksize = 60))
```



## Representing my multiple sequence alingment as an R plot  
using ggmsa package (and function) to plot my virus_alignment... not the one that has been adjusted for seqinR, only the original alignment will allow for this to work correctly:
```{r}
#ggmsa::ggmsa(virus_align) 
```


## Saving my MSA as a PDF
```{r}
#install.package("tinytex")
#install_tinytex
#msaPrettyPrint(virus_align, file="./virus_msa.pdf", askForOverwrite=FALSE)
```

## Creating a genetic distance matrix
Genetic distance is another way of saying evolutionary distance, or how closely two organisms are related. Phylogenetic trees visually explain genetic distance. 
In simple terms, the more alike the genomes are (more nucleotides that are the same, the less genetic distance there is between them and the less evolution has occured)

I will calculate the genetic distance of my viral genomes from eachother using the seqinr dist.alignment() function. 
```{r}
#virus_dist <- seqinr::dist.alignment(virus_align_seqinr, 
                                      # matrix = "identity")
```
an object of class 'dist' is produced

```{r}
#is(virus_dist)
#class(virus_dist)
```

The matrix will be easier to view if it is rounded off... we can use the round function for that
```{r}
#virus_dist_rounded <- round(virus_dist, digits=3)
#virus_dist_rounded
```

## Building a phylogenetic tree.... Finally!
There are many methods of building a phylogenetic tree, I will use the neighbor joininng algorithm (nj()). This will use genetic distances to cluster sequences into clades

I will not use rounded values to build by phylogenetic tree
```{r}
#tree<- nj(virus_dist)
```

#Plotting The Phylogenetic Tree
Using the plot.phylo() function to create the phylogenetic tree, and adding a label
```{r}
#plot.phylo(tree, main="Phylogenetic Tree", 
        #   type="unrooted", 
          # us.edge.length=F)
#mtext(text="Monongevirales order genome tree - UNrooted, no branch lengths")
```

unrooted trees have no groups defined, and the branch lengths have no meanings. 

to make a rooted tree we would remove type = "unrooted" 

```{r}
#plot.phylo(tree, main="Phylogenetic Tree", 
            #use.edge.length = F)


#mtext(text = "Monongevirales order gene tree - rooted, no branch lenths")
```

The tree is now rooted with groups defined, but the branch lengths still have absolutely no meaning. 

We can include information about branch length by using use.edge.length = T
```{r}
#plot.phylo(tree, main="Phylogenetic Tree", 
           # use.edge.length = T)

# add label
#mtext(text = "Monongevirales order genome tree - rooted, with branch lenths")

```
The length of branches now reflect evolutionary distances/relationships between the viruses. 

**only the vertical lines have meaning**


